YouTube Playlist & Videos catalog in Port (#1)
e-ian/port-doc: youtube_playlist_data.md

This guide shows you how to set up a system for tracking YouTube playlists and video performance in Port. 
You’ll learn how to organize your YouTube content, automate the tracking of video metrics, and create visualizations 
to monitor playlist growth and audience engagement. The integration provides centralized content management, 
real-time performance updates, and automated daily data updates, giving you the insights needed to make informed content decisions.

## Prerequisites

Before starting this integration, ensure you have:
   - Sign up at [app.getport.io](https://app.getport.io) and verify your email
   - Log into your Port account at [app.getport.io](https://app.getport.io)
   - Install Port's Github app by clicking [here](https://github.com/apps/getport-io/installations/new)
   - Obtain a YouTube Data API v3 key via the [Google Cloud Console](https://console.cloud.google.com)
   - Prepare your Port organization's Client ID and Client Secret. To find you Port credentials, click [here](https://docs.getport.io/build-your-software-catalog/custom-integration/api/#find-your-port-credentials).

## Data Modeling in Port

### Creating the Playlist Blueprint
   - Log into Port

>    - Click "Builder" in the left sidebar

Reviewer: Add the link to the Builder page

   - Click "+ New Blueprint"
   - In the model that pops up, click on "{---} Edit JSON"
   - Copy and paste the json config below to have for the blueprint properties

<details>
<summary>Click to expand Playlist Blueprint JSON</summary>

> <div class="code-container">

Reviewer: You don't need to write this html. the detail and summary tag will make it a collapsible and If you want it in code just put ```json showLineNumbers  ``` will be fine

<button class="copy-button" onclick="copyCode(this)">Copy</button>

```json
{
  "identifier": "playlist",
  "description": "Youtube playlist description",
  "title": "playlist",
  "icon": "Microservice",
  "schema": {
    "properties": {
      "title": {
        "type": "string",
        "title": "title",
        "description": "title of the playlist"
      },
      "description": {
        "type": "string",
        "title": "description",
        "description": "the description of the playlist"
      },
      "thumbnail_url": {
        "type": "string",
        "title": "thumbnailUrl",
        "description": "the URL of the playlist's thumbnail image",
        "format": "url"
      },
      "video_count": {
        "type": "number",
        "title": "videoCount",
        "description": "The number of videos in the playlist"
      }
    },
    "required": ["title"]
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {}
}
```

> </div>

Reviewer: remove all html please. it's a markdown

  <style>
  .code-container {
    position: relative;
  }

  .copy-button {
    position: absolute;
    top: 5px;
    right: 5px;
    padding: 5px 10px;
    background-color: #f1f1f1;
    border: 1px solid #ccc;
    border-radius: 3px;
    cursor: pointer;
  }

  .copy-button:hover {
    background-color: #e1e1e1;
  }
  </style>

  <script>
  function copyCode(button) {
    const codeBlock = button.nextElementSibling;
    const code = codeBlock.innerText;
    
    navigator.clipboard.writeText(code).then(() => {
      const originalText = button.innerText;
      button.innerText = 'Copied!';
      setTimeout(() => {
        button.innerText = originalText;
      }, 2000);
    }).catch(err => {
      console.error('Failed to copy:', err);
    });
  }
  </script>

</details>

### Creating the Video Blueprint

>    - Click "Builder" in the left sidebar

Reviewer: you can separate this content and let it be under the same header as the data modeling that talks about setting up the blueprints. so you avoid repetition

      },
      "like_count": {
        "type": "number",
        "title": "likeCount",
        "description": "The number of likes the video has received"
      },
      "comment_count": {
        "type": "number",
        "title": "commentCount",
        "description": "The number of comments the video has received"
      }
    },
    "required": ["title"]
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {
    "belongs_to": {
      "title": "Belongs To",
      "description": "relationship between video and playlist",
      "target": "playlist",
      "required": true,
      "many": false
    }
  }
}
```

</div>

  <style>
  .code-container {
    position: relative;
  }

  .copy-button {
    position: absolute;
    top: 5px;
    right: 5px;
    padding: 5px 10px;
    background-color: #f1f1f1;
    border: 1px solid #ccc;
    border-radius: 3px;
    cursor: pointer;
  }

  .copy-button:hover {
    background-color: #e1e1e1;
  }
  </style>

  <script>
  function copyCode(button) {
    const codeBlock = button.nextElementSibling;
    const code = codeBlock.innerText;
    

>     navigator.clipboard.writeText(code).then(() => {

Reviewer: this will be done for you we don't need to write this code

        "playlist_id"
      ],
      "order": [
        "playlist_id"
      ]
    },
    "blueprintIdentifier": "playlist"
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "<YOUR_ORGANISATION_NAME>",
    "repo": "<YOUR_REPO_NAME>",
    "workflow": "<SPECIFY_WORKFLOW_FILE>",
    "workflowInputs": {
      "{{ spreadValue() }}": "{{ .inputs }}",
      "port_context": {
        "runId": "{{ .run.id }}",
        "blueprint": "{{ .action.blueprint }}"
      }
    },
    "reportWorkflowStatus": true
  },
  "requiredApproval": false
}

  ```

  </div>

  <style>
  .code-container {
    position: relative;
  }

  .copy-button {
    position: absolute;
    top: 5px;
    right: 5px;
    padding: 5px 10px;
    background-color: #f1f1f1;
    border: 1px solid #ccc;
    border-radius: 3px;
    cursor: pointer;
  }

  .copy-button:hover {
    background-color: #e1e1e1;
  }
  </style>

  <script>
  function copyCode(button) {
    const codeBlock = button.nextElementSibling;
    const code = codeBlock.innerText;
    
    navigator.clipboard.writeText(code).then(() => {
      const originalText = button.innerText;
      button.innerText = 'Copied!';
      setTimeout(() => {

>         button.innerText = originalText;

Reviewer: remove this

      }, 2000);
    }).catch(err => {
      console.error('Failed to copy:', err);
    });
  }
  </script>

  </details>

> ## GitHub Workflow Setup

Reviewer: ```suggestion
## GitHub workflow 
```

   - Go to your GitHub repository settings
   - Add each of these secrets:
     ```
     YOUTUBE_API_KEY (Your YouTube Data API key)
     PORT_API_KEY (Your Port API key)
     PORT_CLIENT_ID (Your Port Client ID)
     PORT_CLIENT_SECRET (Your Port Client Secret)
     ```

> #### Creating the Workflow File

Reviewer: remove this

>    - Create new file: `ingest.yml` in the directory `.github/workflows`

Reviewer: put this context under workflow and let the needed access keys be an admonition or something

   - Add the following GitHub Actions workflow configuration:

<details>
<summary>Click to expand GitHub Workflow YAML 📋</summary>

>   ```yaml copy

Reviewer: this copy ?? not needed

# YouTube Playlist & Videos catalogue in Port

A comprehensive guide for creating YouTube playlist & Videos catalogue in Port, including data modeling, ingestion, and visualization.

## Overview

> This guide shows you how to set up a system for tracking YouTube playlists and video performance in Port. 
> You’ll learn how to organize your YouTube content, automate the tracking of video metrics, and create visualizations 
> to monitor playlist growth and audience engagement. The integration provides centralized content management, 
> real-time performance updates, and automated daily data updates, giving you the insights needed to make informed content decisions.

Reviewer 2: ```suggestion
This guide shows you how to set up a system for tracking YouTube playlists and video performance in Port. 
You’ll learn how to organize your YouTube content, automate the tracking of video metrics, and create visualizations  to monitor playlist growth and audience engagement.
```

## Prerequisites

Before starting this integration, ensure you have:
   - Sign up at [app.getport.io](https://app.getport.io)

>    - Log into your Port account at [app.getport.io](https://app.getport.io)

Reviewer 2: this can be merged into 1

   - Install Port's Github app by clicking [here](https://github.com/apps/getport-io/installations/new)
   - Obtain a YouTube Data API v3 key via the [Google Cloud Console](https://console.cloud.google.com)
   - Prepare your Port organization's Client ID and Client Secret. To find you Port credentials, click [here](https://docs.getport.io/build-your-software-catalog/custom-integration/api/#find-your-port-credentials).

## Data Modeling in Port
This section outlines the data model for managing YouTube content in Port. The model consists of two main blueprints: `playlist` and `video`, designed to track YouTube playlists and their associated videos. The relationship between playlists and videos is maintained through a one-to-many relationship, where each video belongs to a playlist.

### Setting up Blueprints
   - Log into Port
   - Click ["Builder"](https://app.getport.io/settings/data-model) in the left sidebar
   - Click "+ New Blueprint"
   - In the model that pops up, click on "{---} Edit JSON"
   - Copy and paste the json configs below to have for the Playlist and Video blueprint properties

<details>

> <summary>Click to expand Playlist Blueprint JSON</summary>

Reviewer: Make the other ones like this as well please. 
And use capitalization in the titles rather than making all of them start with uppercase, let the first word be the only one starting with uppercase

  "identifier": "playlist",
  "description": "Youtube playlist description",
  "title": "YouTube Playlist",
  "icon": "Youtrack",
  "schema": {
    "properties": {
      "description": {
        "type": "string",

>         "title": "description",

Reviewer 2: ```suggestion
        "title": "Description",
```

>   "icon": "Microservice",

Reviewer 2: change to a better icon. and rename title to  YouTube Playlist

        "description": "the description of the playlist"
      },
      "thumbnail_url": {
        "type": "string",

>         "title": "thumbnailUrl",

Reviewer 2: ```suggestion
        "title": "Thumbnail Url",
```

>       },

Reviewer 2: let's delete this property since every entity is required to have one. so adding this will be redundant

        "description": "the URL of the playlist's thumbnail image",
        "format": "url"
      },
      "video_count": {
        "type": "number",

>         "title": "videoCount",

Reviewer 2: ```suggestion
        "title": "Video Count",
```

        "type": "string",
        "title": "thumbnailUrl",
        "description": "the URL of the playlist's thumbnail image",
        "format": "url"
      },
      "video_count": {
        "type": "number",
        "title": "videoCount",
        "description": "The number of videos in the playlist"
      }
    },

>     "required": ["title"]

Reviewer 2: title should be removed

<details>
<summary><b> Video blueprint (click to expand) </b></summary>

```json
{
  "identifier": "video",
  "description": "youtube video blueprint",
  "title": "YouTube Video",
  "icon": "Youtrack",
  "schema": {
    "properties": {
      "description": {
        "type": "string",

>         "title": "description",

Reviewer 2: ```suggestion
        "title": "Description",
```

> <summary>Click to expand Video Blueprint JSON</summary>

Reviewer: ```suggestion
<summary><b> Video blueprint (click to expand) </b></summary>
```

        "description": "the description of the video"
      },
      "thumbnail_url": {
        "type": "string",

>         "title": "thumbnailUrl",

Reviewer 2: ```suggestion
        "title": "Thumbnail Url",
```

>   "title": "video",

Reviewer 2: should be capitalized and renamed to YouTube Video

>   "icon": "Microservice",

Reviewer 2: change to a better icon other than microservice

    "properties": {
      "title": {
        "type": "string",
        "title": "title",
        "description": "the title of the video"

>       },

Reviewer 2: delete this since it's redundant

        "description": "the duration of the video"
      },
      "view_count": {
        "type": "number",
        "title": "viewCount",
        "description": "The number of views the video has received"
      },
      "like_count": {
        "type": "number",
        "title": "likeCount",
        "description": "The number of likes the video has received"
      },
      "comment_count": {
        "type": "number",
        "title": "commentCount",
        "description": "The number of comments the video has received"
      }

>     }

Reviewer 2: just like I've suggested above, the title of the properties should in this format : Video Count, Comment Count, Duration, Description etc

  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {
    "belongs_to": {

>       "title": "playlist",

Reviewer 2: ```suggestion
        "title": "Playlist",
```

      },
      "comment_count": {
        "type": "number",
        "title": "commentCount",
        "description": "The number of comments the video has received"
      }
    },
    "required": ["title"]
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {
    "belongs_to": {

>       "title": "Belongs To",

Reviewer: you shouldn't name relationships like this. 
The name of the relationship should have a meaning. this is on the video blueprint so it has a relationship with the playlist. change the name to playlist

      "description": "relationship between video and playlist",
      "target": "playlist",

>       "required": true,

Reviewer 2: this should be set to false because not every video belongs to a playlist

      "many": false
    }
  }
}
```

</details>

> ## Self service setup

Reviewer 2: You can add a one liner explaining why we need a self service action (SSA) for this task

  - Navigate to Self-service in Port by clicking [here](https://app.getport.io/self-serve)

>   - Click on the "+ Action" button to create and new action. Click on "{---} Edit JSON" in the pop up

Reviewer: ```suggestion
  - Click on the "+ Action" button to create and new action. 
  - Click on "{---} Edit JSON" in the pop up
```

Reviewer: Make these steps as well

>   - Copy and paste the json configuration details below to setup the Self-Service. Edit with your details.

Reviewer: ```suggestion
  - Copy and paste the json configuration details below to setup the Self-Service. 
  :::tip Placeholders
  Remember to replace <YOUR_ORGANISATION_NAME> with your organization name
  <YOUR_REPO_NAME> with your repository name and <SPECIFY_WORKFLOW_FILE> with the name of 
  the workflow
  :::
```

  <details>

>   <summary>Click to expand Self-service config JSON</summary>

Reviewer: ```suggestion
  <summary><b> Action configuration (click to expand) </b></summary>
```

  "trigger": {
    "type": "self-service",
    "operation": "CREATE",
    "userInputs": {
      "properties": {
        "playlist_id": {
          "icon": "Youtrack",
          "type": "string",

>           "title": "playlist_id",

Reviewer 2: ```suggestion
        "title": "Playlist ID",
```

    "operation": "CREATE",
    "userInputs": {
      "properties": {
        "playlist_id": {
          "icon": "Youtrack",
          "type": "string",
          "title": "playlist_id",
          "description": "Playlist id to be used for fetching the different respective videos"
        }
      },
      "required": [
        "playlist_id"
      ],
      "order": [
        "playlist_id"
      ]
    },
    "blueprintIdentifier": "playlist"
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "<YOUR_ORGANISATION_NAME>",
    "repo": "<YOUR_REPO_NAME>",
    "workflow": "<SPECIFY_WORKFLOW_FILE>",
    "workflowInputs": {
      "{{ spreadValue() }}": "{{ .inputs }}",
      "port_context": {
        "runId": "{{ .run.id }}",
        "blueprint": "{{ .action.blueprint }}"
      }
    },
    "reportWorkflowStatus": true
  },
  "requiredApproval": false
}

  ```

  </details>

## GitHub workflow

>    - Go to your GitHub repository settings

Reviewer: make this steps as well

   - Add each of these secrets:

>      ```

Reviewer: ```suggestion
    
```

>      YOUTUBE_API_KEY (Your YouTube Data API key)

Reviewer: ```suggestion
     - YOUTUBE_API_KEY (Your YouTube Data API key)
     - PORT_API_KEY (Your Port API key)
     - PORT_CLIENT_ID (Your Port Client ID)
     - PORT_CLIENT_SECRET (Your Port Client Secret)
```

>      PORT_API_KEY (Your Port API key)

Reviewer: ```suggestion
    
```

>      PORT_CLIENT_ID (Your Port Client ID)

Reviewer: ```suggestion
     
```

>      PORT_CLIENT_SECRET (Your Port Client Secret)

Reviewer: ```suggestion
     
```

>      ```

Reviewer: ```suggestion
     
```

>    - Create new file: `ingest.yml` in the directory `.github/workflows` and pass the secrets.

Reviewer 2: rephrase this line. It make readers believe that they need to pass the secret to this file

>    - Add the following GitHub Actions workflow configuration:

Reviewer 2: Github Actions workflow is redundant. Could say GitHub workflow

            PORT_CONTEXT: ${{ inputs.port_context }}
          run: |
            set -e
            echo "::group::Fetching playlist data"
            PLAYLIST_DATA=$(curl -s "https://youtube.googleapis.com/youtube/v3/playlists?part=snippet,contentDetails&id=${PLAYLIST_ID}&key=${YOUTUBE_API_KEY}")
            
            if [ "$(echo $PLAYLIST_DATA | jq '.items | length')" -eq 0 ]; then
              echo "::error::No playlist found"
              exit 1
            fi

            TITLE=$(echo $PLAYLIST_DATA | jq -r '.items[0].snippet.title')
            DESC=$(echo $PLAYLIST_DATA | jq -r '.items[0].snippet.description')
            THUMB=$(echo $PLAYLIST_DATA | jq -r '.items[0].snippet.thumbnails.default.url')
            COUNT=$(echo $PLAYLIST_DATA | jq -r '.items[0].contentDetails.itemCount')

            PLAYLIST_PAYLOAD=$(jq -n \
              --arg id "$PLAYLIST_ID" \
              --arg title "$TITLE" \
              --arg desc "$DESC" \
              --arg thumb "$THUMB" \
              --arg count "$COUNT" \
              '{
                identifier: $id,
                title: $title,
                properties: {

>                   title: $title,

Reviewer 2: remove title

                  local V_THUMB=$(echo "$VIDEO_DATA" | jq -r '.items[0].snippet.thumbnails.default.url')
                  local V_DURATION=$(echo "$VIDEO_DATA" | jq -r '.items[0].contentDetails.duration')
                  local V_VIEWS=$(echo "$VIDEO_DATA" | jq -r '.items[0].statistics.viewCount // "0"')
                  local V_LIKES=$(echo "$VIDEO_DATA" | jq -r '.items[0].statistics.likeCount // "0"')
                  local V_COMMENTS=$(echo "$VIDEO_DATA" | jq -r '.items[0].statistics.commentCount // "0"')

                  add_action_log "Found video: $V_TITLE" "Found"

                  VIDEO_PAYLOAD=$(jq -n \
                    --arg id "$VIDEO_ID" \
                    --arg title "$V_TITLE" \
                    --arg desc "$V_DESC" \
                    --arg thumb "$V_THUMB" \
                    --arg duration "$V_DURATION" \
                    --arg views "$V_VIEWS" \
                    --arg likes "$V_LIKES" \
                    --arg comments "$V_COMMENTS" \
                    --arg playlist_id "$PLAYLIST_ID" \
                    '{
                      identifier: $id,
                      title: $title,
                      properties: {

>                         title: $title,

Reviewer 2: remove title

            PROCESSED=$(cat /tmp/videos_processed)
            FAILED=$(cat /tmp/videos_failed)
            
            FINAL_DETAILS="Processed ${PROCESSED} videos, ${FAILED} failed"
            add_action_log "$FINAL_DETAILS" "Completed"
            
            if [ "$PROCESSED" -gt 0 ]; then
              update_final_status "SUCCESS" "Processing complete" "$FINAL_DETAILS"
            else
              update_final_status "FAILURE" "No videos processed" "$FINAL_DETAILS"
              exit 1
            fi

        - name: Report Failure
          if: failure()
          env:
            PORT_CONTEXT: ${{ inputs.port_context }}
          run: |
            set -e
            RUN_ID=$(echo "$PORT_CONTEXT" | jq -r --raw-input 'fromjson | .runId')
            
            curl -s -X PATCH "https://api.getport.io/v1/actions/runs/$RUN_ID" \
              -H "Authorization: Bearer $ACCESS_TOKEN" \
              -H "Content-Type: application/json" \
              -d '{
                "status": "FAILURE",
                "message": {
                  "summary": "Workflow failed",
                  "details": "Check logs for details"
                }
              }'

> 

Reviewer 2: can we also report success to Port?

  ```

</details>

## Example self-service action
- On the Self-service tab, navigate to the self-service you created and click on "Create".
- The pop-up below will be shown, enter the value of the youtube `playlist_id` and click execute
<details>
<summary>Click to see example Self-Service Execution</summary>
<img src="./assets/execute.png" alt="Self-Service Execution">
</details>
- This will trigger the workflow and you can track the progress from your port account.

> - When completed, go to catalogue to view the playlist and videos respectively.

Reviewer 2: catalog instead of catalogue

## Creating Visualizations

After the data is ingested, create these visualizations to monitor your YouTube content:

> ### 1. Video Count Metric

Reviewer 2: Can you also show the sections where you create the dashboard? what you've shown so far is the dashboard widgets

3. Select "Number Chart"
4. Configure:

<details>
<summary>Click to see details and example configuration image</summary>
<img src="./assets/videocount.png" alt="Video count in playlist">
</details>

### 2. Average Likes Metric
1. Click "+ Add Widget"
2. Select "Number Chart"
3. Configure:

<details>
<summary>Click to see details and example configuration image</summary>
<img src="./assets/averagelikes.png" alt="Average likes card">
</details>

### 3. Video Details Table
1. Click "+ Add Widget"
2. Select "Table"
3. Configure:

<details>
<summary>Click to see details and example configuration image</summary>
<img src="./assets/videodetails.png" alt="Video details table">
</details>

### 4. Engagement Distribution Pie Chart
1. Click "+ Add Widget"
2. Select "Pie Chart"
3. Configure:

<details>
<summary>Click to see details and example configuration image</summary>
<img src="./assets/engagementdist.png" alt="Video engagement distribution">
</details>

### YouTube Playlist Analytics Dashboard Layout

<details>
<summary>Click to see example YouTube Analytics Dashboard</summary>
<img src="./assets/youtube_analytics_dashboard.png" alt="YouTube Analytics Dashboard">

> </details>

Reviewer 2: The guide is missing conclusion section

# YouTube Playlist & Videos catalogue in Port

A comprehensive guide for creating YouTube playlist & Videos catalogue in Port, including data modeling, ingestion, and visualization.

> ## Table of Contents

Reviewer: We don't need table of content. please look at sample guides.

- [Overview](#overview)
- [Prerequisites](#prerequisites)
- [Part 1: Data Modeling in Port](#part-1-data-modeling-in-port)
  - [Creating the Playlist Blueprint](#creating-the-playlist-blueprint)
  - [Creating the Video Blueprint](#creating-the-video-blueprint)
  - [Blueprint YAML Representations](#blueprint-yaml-representations)
- [Part 2: GitHub Workflow Setup](#part-2-github-workflow-setup)
  - [Setting Up Repository Secrets](#setting-up-repository-secrets)
  - [Creating the Workflow File](#creating-the-workflow-file)
- [Part 3: Data Ingestion](#part-3-data-ingestion)
- [Part 4: Creating Visualizations](#part-4-creating-visualizations)
- [Value Proposition](#value-proposition)
- [Troubleshooting](#troubleshooting)
- [Relevant Links](#relevant-links)
- [Author](#author)

## Overview

This guide demonstrates how to create a comprehensive YouTube playlist tracking system in Port. You'll learn how to:

> - Model playlist and video data using Port blueprints

Reviewer: overview shouldn't be bulleted, let's have an introduction statement

- Set up automated data ingestion using GitHub Actions
- Create meaningful visualizations of your YouTube data
- Monitor playlist and video performance metrics

> ## Prerequisites

Reviewer: Again check out sample guide on Port. Prerequisites should be simplified and with this particular assessment you don't need a lot of these

Before starting this integration, ensure you have:

1. **Port Account**
   - Sign up at [app.getport.io](https://app.getport.io)
   - Verify your email and log in

   **Installing Port's GitHub App**
   - Log into your Port account at [app.getport.io](https://app.getport.io)
   - Navigate to "Data Sources" in the left sidebar
   - Find and click on "GitHub" under the available integrations
   - Click "Install GitHub App"
   - You will be redirected to GitHub's app installation page
   - Choose whether to install the app on all repositories or select specific ones
   - If selecting specific repositories:
     - Choose the repositories you want to monitor
     - Click "Install & Authorize"
   - Return to Port's "Data Sources" page
   - Check that GitHub appears as a connected integration
   - The status should show as "Active"

> 

Reviewer: again simplify it

2. **GitHub Account**
   - Active GitHub account
   - Repository where you'll set up the workflow

3. **API Keys and Credentials**
   - YouTube Data API v3 key
     - Go to [Google Cloud Console](https://console.cloud.google.com)
     - Create a project or select existing one
     - Enable YouTube Data API v3
     - Create API credentials
   - Port API key
   - Port Client ID and Secret

## Part 1: Data Modeling in Port

> ### Creating the Playlist Blueprint

Reviewer: talk about the navigation to the page and have a json they can just copy and paste. configure the json blueprint to have your properties

   - Type: "String"
   - Format: "URL"
   - Required: No
   - Description: "The URL of the video's thumbnail image"
   
   d) Duration Property
   - Name: "duration"
   - Type: "String"
   - Required: No
   - Description: "The duration of the video"
   
   e) View Count Property
   - Name: "view_count"
   - Type: "Number"
   - Required: No
   - Description: "The number of views the video has received"
   
   f) Like Count Property
   - Name: "like_count"
   - Type: "Number"
   - Required: No
   - Description: "The number of likes the video has received"
   
   g) Comment Count Property
   - Name: "comment_count"
   - Type: "Number"
   - Required: No
   - Description: "The number of comments the video has received"

3. **Add Relation**
   - Click "Add Relation"
   - Name: "belongs_to"
   - Title: "Belongs To"
   - Description: "Relationship between video and playlist"
   - Target Blueprint: Select "playlist"
   - Required: Yes
   - Many: No

4. **Save Blueprint**
   - Click "Create"
   - Verify the blueprint appears with all properties and relations

### Blueprint YAML Representations

When creating the blueprints in Port, you can also use these YAML definitions:

> #### Playlist Blueprint YAML

Reviewer: the blueprint is not a yaml file. it's a JSON file please check out the docs well

```yaml

> identifier: playlist

Reviewer: same as previous comment

description: "Youtube playlist description"
title: "playlist"
icon: "Microservice"
schema:
  properties:
    title:
      type: "string"
      title: "title"
      description: "title of the playlist"
    description:
      type: "string"
      title: "description"
      description: "the description of the playlist"
    thumbnail_url:
      type: "string"
      title: "thumbnailUrl"
      description: "the URL of the playlist's thumbnail image"
      format: "url"
    video_count:
      type: "number"
      title: "videoCount"
      description: "The number of videos in the playlist"
  required:
    - "title"
mirrorProperties: {}
calculationProperties: {}
aggregationProperties: {}
relations: {}
```

> #### Video Blueprint YAML

Reviewer: same here

                            playlist_items_request, playlist_items_response)
                        
                except Exception as e:
                    logger.error(f"Error processing playlist videos: {str(e)}")
                    raise

            def run(self):
                try:
                    # First update/create playlist entity
                    playlist_data = self.fetch_playlist_info()
                    self.update_or_create_entity(playlist_data, self.playlist_blueprint_identifier)
                    
                    # Then process all videos in the playlist
                    self.process_playlist_videos()
                    
                    logger.info("Successfully completed YouTube Port ingestion")
                    
                except Exception as e:
                    logger.error(f"Error during ingestion run: {str(e)}")
                    raise

        if __name__ == "__main__":
            try:
                ingestion = YouTubePortIngestion()
                ingestion.run()
            except Exception as e:
                logger.error(f"Failed to complete ingestion: {str(e)}")
                sys.exit(1)
        EOF

> ```

Reviewer: since you had the file written already. you should have called it here in the md.  change this

3. **Important Configuration Notes**
   - Replace `PLAYLIST_ID` in the code with your actual YouTube playlist ID
   - The workflow runs daily at midnight UTC (`cron: "0 0 * * *"`)
   - Manual triggering is enabled through `workflow_dispatch`

>    - The script includes retry logic and rate limiting for API calls

Reviewer: the automation should be done in port like I mentioned

   - Comprehensive error handling and logging is implemented

> 4. **Rate Limiting and Quotas**

Reviewer: we don't need to implement rate limiting in this. it's not an integration. it's a self service action or an automation

   - The script includes a 0.1-second delay between video processing
   - YouTube API has daily quota limits - monitor your usage
   - Port API calls include retry logic for better reliability

## Value Proposition

Implementing this YouTube playlist integration with Port provides several key benefits:

> 

Reviewer: value preposition should be part of your overview contest. we don't need a separate header for it

   - Analyze engagement patterns
   - Make informed decisions about content strategy

4. **Automated Tracking**
   - Daily automatic updates of video metrics
   - No manual data entry required
   - Always up-to-date analytics

## Part 4: Creating Visualizations

After the data is ingested, create these visualizations to monitor your YouTube content:

### 1. Video Count Metric
1. Navigate to "Dashboards" in Port
2. Click "+ Add Widget"
3. Select "Number Chart"
4. Configure:
   - Title: "Total Videos"
   - Under Query Definition:
     - Blueprint: Select "video"
     - Aggregation: Select "Count"

### 2. Average Likes Metric
1. Click "+ Add Widget"
2. Select "Number Chart"
3. Configure:
   - Title: "Average Likes"
   - Under Query Definition:
     - Blueprint: Select "video"
     - Aggregation: Select "Average"
     - Property: Select "like_count"

### 3. Video Details Table
1. Click "+ Add Widget"
2. Select "Table"
3. Configure:
   - Title: "Video Details"
   - Select Blueprint: "video"
   - Add Columns:
     - Title
     - View Count
     - Like Count
     - Comment Count
     - Duration

### 4. Engagement Distribution Pie Chart
1. Click "+ Add Widget"
2. Select "Pie Chart"

> 3. Configure:

Reviewer: the steps should help the client replicate exactly what you've created we will be adding pictures to make this work best. check out sample guides

   - Title: "Video Engagement Distribution"
   - Select Blueprint: "video"
   - Group By: "view_count"

### YouTube Playlist Analytics Dashboard Layout

![YouTube Analytics Dashboard](./assets/youtube_analytics_dashboard.png)

> ## Relevant Links

Reviewer: remove this please

> 1. **Port Resources**

Reviewer: remove this too, the client doesn't need it

   - [Port Documentation](https://docs.getport.io/)
   - [Port API Reference](https://api-docs.getport.io/)
   - [Port Blueprints Guide](https://docs.getport.io/build-your-software-catalog/define-your-data-model/setup-blueprint/)

2. **YouTube Resources**
   - [YouTube Data API Documentation](https://developers.google.com/youtube/v3)
   - [YouTube API Quotas and Compliance](https://developers.google.com/youtube/v3/getting-started#quota)

3. **GitHub Resources**
   - [GitHub Actions Documentation](https://docs.github.com/en/actions)
   - [GitHub Secrets Management](https://docs.github.com/en/actions/security-guides/encrypted-secrets)
   - [GitHub Repo link](https://github.com/e-ian/port-doc/blob/main/youtube_playlist_data.md)
   

## Troubleshooting

> Common issues and their solutions:

Reviewer: And this should be removed too

1. **Workflow Fails to Run**
   - Check if all secrets are properly set in GitHub
   - Verify YouTube API key has necessary permissions
   - Ensure Port credentials are valid

2. **Data Not Appearing in Port**
   - Verify blueprint definitions match the ingestion code
   - Check workflow logs for any API errors
   - Ensure relations are properly configured

3. **Visualization Issues**
   - Refresh the dashboard
   - Verify data is present in Port
   - Check widget configurations

> ## Author

Reviewer: This too should be removed
